### ISSUE TITLE
Indicate which sections of a method do not return, if possible

### ISSUE THREAD

**Issue by Pilchard123** [dotnet/roslyn-analyzers#4493]
### Analyzer

**Diagnostic ID**: [CS0161](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0161)

### Describe the improvement

It may be useful if the message for CS0161 could show which branches in a method do not return. In some cases it is not immediately obvious why the method does not return, either because the method has convoluted branching or because the writer has misunderstood or mistyped a condition and so mistakenly believes that all cases are covered. This proposed change would indicate which particular section is likely to be incorrect.

### Describe suggestions on how to achieve the rule

Perhaps given the code in the context section below the message could say something similar to "not all code paths return a value - lines A-B, Z-end", with A-B being a section within a method but not at the end and Z being the start of a section that ends a method.

This may be better done as a new analyzer, with one report for each section that if entered will never return or throw, with messages like "Section from lines X-Y will never return" or "Method will never return if line Z is reached".

### Additional context

```csharp
int Foo()
{
    if(condition)
    {
        return 1;
    }
    else
    { // A
        
    } // B
}

int Bar()
{
    if(condition)
    {
        return 1;
    }
    // Z
}
```

**Comment by mavasani** [comment]
I believe this is already covered by https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0035. When you open a source with unreachable code blocks in Visual studio, you should see an IDE0035 for each such section and the section is also faded out. Additionally, there is a code fix to remove unreachable block.

**Comment by Pilchard123** [comment]
Not quite - this is for code that exists and can be reached, but that will never return. Consider the toy function `Bar`, as above:

```csharp
int Bar()
{
    if (condition)
    {
        return 1;
    }
    // we get to here if condition is false
    // but if condition *is* false, then we get to a path that has no return
}
```

We already have something similar in CS8509. If you have something like

```csharp
var bool1 = true;
var bool2 = false;

var foo = (bool1, bool2) switch
{
    (true, true) => true,
    (true, false) => false,
    (false, true) => false
};
```

then you will get "The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern '(false, false)' is not covered." I'm proposing a similar analyzer but for methods.

**Comment by mavasani** [comment]
The compiler should already generate an error when a code path has a missing return. Can you please clarify the request?

**Comment by Pilchard123** [comment]
The will currently generate an error if there is a path that does not return. I'm asking if it's possible to identify _which_ path(s) don't return.

**Comment by Pilchard123** [comment]
As an example, the code below gives the error "'Program.Main()': not all code paths return a value". Could it instead (or as well) say something like "'Program.Main()' will not return a value if line 16 is reached"?

```
using System;

namespace IssueDemo
{
    class Program
    {
        static int Main()
        {
            var rng = new Random();
            if (rng.Next() % 2 == 0)
            {
                return 1;
            }
            else
            {

            }
        }
    }
}
```

**Comment by Youssef1313** [comment]
@mavasani Looks like this is a suggestion for compiler diagnostic message. Can you move to dotnet/roslyn?
